server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -595,51 +627,67 @@ private void deleteIndices(RepositoryData repositoryData, List<IndexId> indices,
         final Executor executor = threadPool.executor(ThreadPool.Names.SNAPSHOT);
         for (IndexId indexId : indices) {
-            executor.execute(ActionRunnable.wrap(deleteIndexMetaDataListener,
-                deleteIdxMetaListener -> {
-                    final IndexMetaData indexMetaData;
-                    try {
-                        indexMetaData = getSnapshotIndexMetaData(snapshotId, indexId);
-                    } catch (Exception ex) {
-                        logger.warn(() ->
-                            new ParameterizedMessage("[{}] [{}] failed to read metadata for index", snapshotId, indexId.getName()), ex);
-                        // Just invoke the listener without any shard generations to count it down, this index will be cleaned up
-                        // by the stale data cleanup in the end.
-                        deleteIdxMetaListener.onResponse(null);
-                        return;
-                    }
-                    final int shardCount = indexMetaData.getNumberOfShards();
-                    assert shardCount > 0 : "index did not have positive shard count, get [" + shardCount + "]";
-                    // Listener for collecting the results of removing the snapshot from each shard's metadata in the current index
-                    final ActionListener<ShardSnapshotMetaDeleteResult> allShardsListener =
-                        new GroupedActionListener<>(deleteIdxMetaListener, shardCount);
-                    final Index index = indexMetaData.getIndex();
-                    for (int shardId = 0; shardId < indexMetaData.getNumberOfShards(); shardId++) {
-                        final ShardId shard = new ShardId(index, shardId);
-                        executor.execute(new AbstractRunnable() {
-                            @Override
-                            protected void doRun() throws Exception {
-                                allShardsListener.onResponse(
-                                    deleteShardSnapshot(repositoryData, indexId, shard, snapshotId));
+            final Set<SnapshotId> survivingSnapshots = repositoryData.getSnapshots(indexId).stream()
+                .filter(id -> id.equals(snapshotId) == false).collect(Collectors.toSet());
+            executor.execute(ActionRunnable.wrap(deleteIndexMetaDataListener, deleteIdxMetaListener -> {
+                final IndexMetaData indexMetaData;
+                try {
+                    indexMetaData = getSnapshotIndexMetaData(snapshotId, indexId);
+                } catch (Exception ex) {
+                    logger.warn(() ->
+                        new ParameterizedMessage("[{}] [{}] failed to read metadata for index", snapshotId, indexId.getName()), ex);
+                    // Just invoke the listener without any shard generations to count it down, this index will be cleaned up
+                    // by the stale data cleanup in the end.
+                    deleteIdxMetaListener.onResponse(null);
```

Comment:
What does this mean for newer ES versions where we rely on a new IndexShardSnapshot file to be written by this method?

Reply:
yes, I don't have a clear idea yet how this should be handled, but I now think that the current handling is safe. Let's keep as is (perhaps add a TODO).